cabal-install/Distribution/Client/IndexUtils/Timestamp.hs

```diff
@@ -0,0 +1,117 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE RecordWildCards #-}
+
+-----------------------------------------------------------------------------
+-- |
+-- Module      :  Distribution.Client.IndexUtils.Timestamp
+-- Copyright   :  (c) Herbert Valerio Riedel
+-- License     :  BSD3
+--
+-- Timestamp type used in package indexes
+
+module Distribution.Client.IndexUtils.Timestamp
+    ( Timestamp
+    , nullTimestamp
+    , epochTimeToTimestamp
+    , timestampToUTCTime
+    , utcTimeToTimestamp
+    ) where
+
+import qualified Codec.Archive.Tar.Entry    as Tar
+import           Control.DeepSeq
+import           Data.Char                  (isDigit)
+import           Data.Int                   (Int64)
+import           Data.Time                  (UTCTime (..), fromGregorianValid,
+                                             makeTimeOfDayValid, showGregorian,
+                                             timeOfDayToTime, timeToTimeOfDay)
+import           Data.Time.Clock.POSIX      (posixSecondsToUTCTime,
+                                             utcTimeToPOSIXSeconds)
+import           Distribution.Compat.Binary
+import qualified Distribution.Compat.ReadP  as ReadP
+import           Distribution.Text
+import qualified Text.PrettyPrint           as Disp
+
+-- | UNIX timestamp (expressed in seconds since unix epoch, i.e. 1970).
+newtype Timestamp = TS Int64 -- Tar.EpochTime
+                  deriving (Eq,Ord,Enum,NFData)
```

Comment:
fwiw, a later patch adds a derived `Show` instance anyway... I could pull that one into here